### PR TITLE
T/329 Split Selection to Selection and DocumentSelection

### LINKED ISSUE #329: Split Selection into (Static)Selection and LiveSelection
The request comes from the question "what are we going to pass to methods like `deleteContents()`?". In CKEditor 4 we usually had a command that automatically works on the real selection and helper methods which accept ranges. The command of course call one of the helper methods, so they exist mostly for code reusability reasons. Other algorithms can use those in order to build some more complex behaviours.

In CKEditor 5 the 4's scheme with util methods working on ranges does not fully make sense, because:

1. Simple array of ranges doesn't specify a direction. So it would need to be accompanied with another attribute.
1. Selection defines also its attributes, so the array of ranges and direction would need also this.
1. A simple set of ranges do not have selection's method from which some may be useful.

In other words, it's clearly visible that algorithms need to work on selection. Theoretically we could now do this:

```js
const sel = new Selection( doc );

sel.addRange( someRange );

deleteUtils.delete( sel );
```

However, the current implementation of the `Selection` operates on live ranges, what mean that the selection needs to be detached once not needed. Otherwise, the editor will quickly become sluggish due to huge number of attached live positions.

Therefore, let's split the selection into two classes. Basic, static selection that can be used as a feed for utils, and live selection for the document's main selection (and those who know what they are doing).

The split should be pretty simple to do as live ranges are created in just one place. 

PS. For now we can freely live without the static selection. Simply, in the tests (which, for a time being, will be the only code that uses the feature utils) we can use live selection and detach it when needed.

### PR DESCRIPTION
**Pull request by scofalik**
This PR fixes #329 

Notable changes:
* Introduced `DocumentSelection` which has same features as "old" `Selection`.
* `Selection` is now static, meaning that it does not create `LiveRange`s and does not update when document changes.
* `Selection` has no attributes API.
* `Selection` has no "default range".
* `DocumentSelection` has `getSnapshot` method that returns `Selection` instance with same ranges and direction as that `DocumentSelection`.

New `Selection` class is meant to be used when you need to cached it or in methods that may/have to modify selection.

### CODE REVIEW

**Comment by pjasiun** [comment]
The only important difference between `Selection` and `DocumentSelection` is that `DocumentSelection` use `LiveRange` instead of regular ranges. It means that `DocumentSelection` should be called `LiveSelection` for consistency.

**Comment by scofalik** [comment]
Done

File: src/model/document.js [hunk 1]
```diff
@@ -162,7 +162,6 @@ export default class Document {
 		this.version++;
 
 		if ( operation.delta ) {
-			// Right now I can't imagine operations without deltas, but let's be safe.
 			this.history.addDelta( operation.delta );
```
Comment (pjasiun) [inline src/model/document.js]: Can you merge master?

**Comment by scofalik** [comment]
Done.

File: src/model/liveselection.js [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+	}
+
+	/**
+	 * Creates an instance of {@link engine.model.Selection} that has same ranges and direction as this selection. Since
+	 * this will be instance of `Selection` instead of `LiveSelection`, it will not be automatically updated or rendered,
+	 * so it can be used in algorithms using and modifying selection.
+	 *
+	 * @returns {engine.model.Selection} Selection instance which ranges and direction is equal to this selection.
+	 */
+	getSnapshot() {
```
Comment (pjasiun) [inline src/model/liveselection.js]: We have `Position.createFromPosition`, `Range.createFromRange` so why not `Selection.createFromSelection`?

Comment (pjasiun) [inline src/model/liveselection.js]: In fact I do not like these method, I prefer `.clone( isDynamic )` but it is more important to have a consistent API now, we can discuss it in the separate issue.

Comment (scofalik) [inline src/model/liveselection.js]: In fact, why not Zoidberg?

File: src/model/selection.js [hunk 3]
```diff
@@ -60,7 +34,7 @@ export default class Selection {
 		 * Stores all ranges that are selected.
 		 *
 		 * @private
```
Comment (pjasiun) [inline src/model/selection.js]: It should be protected, not private, if `LiveSelection` use it.

File: src/model/liveselection.js (hunk 2 above)
Comment (pjasiun) [inline src/model/liveselection.js]: I created a follow-up: https://github.com/ckeditor/ckeditor5-engine/issues/507

**Comment by scofalik** [comment]
Done.

**Pull request merged**